File: Client/Frontend/Browser/FocusHelper.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0

import Foundation
import Shared
import WebKit

private let log = Logger.browserLogger

class FocusHelper: TabContentScript {
    fileprivate weak var tab: Tab?

    init(tab: Tab) {
        self.tab = tab
    }

    static func name() -> String {
        return "FocusHelper"
    }

    func scriptMessageHandlerName() -> String? {
        return "focusHelper"
    }

    func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
        guard let data = message.body as? [String: String] else {
            return log.error("FocusHelper.js sent wrong type of message")
        }

        guard let _ = data["elementType"],
            let eventType = data["eventType"] else {
            return log.error("FocusHelper.js sent wrong keys for message")
        }

        switch eventType {
        case "focus":
            tab?.isEditing = true
        case "blur":
            tab?.isEditing = false
        default:
            return log.error("FocusHelper.js sent unhandled eventType")
        }
    }
}

File: firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentEnd/FocusHelper.js
/* vim: set ts=2 sts=2 sw=2 et tw=80: */
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/

"use strict";

// Ensure this module only gets included once. This is
// required for user scripts injected into all frames.
window.__firefox__.includeOnce("FocusHelper", function() {
  const isButton = (element) => {
    if (element.nodeName !== "INPUT") {
      return false;
    }

    const type = element.type.toUpperCase();
    return (type == "BUTTON" || type == "SUBMIT" || type == "FILE");
  };

  const handler = (event) => {
    const eventType = event.type;
    const elementType = event.target.nodeName;
    // We can receive focus and blur events from `a` elements and anything with a `tabindex` attribute.
    // We should also not fire for buttons..
    if (elementType === "INPUT" || elementType === "TEXTAREA" || event.target.isContentEditable) {
      if (!isButton(event.target)) {
        webkit.messageHandlers.focusHelper.postMessage({
          eventType,
          elementType
        });
      }
    }
  };

  const options = {
    capture: true,
    passive: true,
  };

  const body = window.document.body;
  // In certain contexts, like PDF documents, the body might not exist, hence the optional chaining.
  body?.addEventListener("focus", handler, options);
  body?.addEventListener("blur", handler, options);
});
